source/dub/generators/generator.d

```diff
@@ -143,144 +145,278 @@ class ProjectGenerator
+			logDebug("deps: %s -> %(%s, %)", pack.name, deps.byKey);
+			foreach (depname; deps.keys.sort())
+			{
+				auto depspec = deps[depname];
+				// get selected package for that dependency, e.g. vibe-d 0.8.2-beta.2
+				auto deppack = m_project.getDependency(depname, depspec.optional);
+				if (deppack is null) continue; // optional and not selected
+
+				// if dependency is not a target itself
+				if (!isBinaryTarget[depname]) {
+					// add itself
+					ti.packages ~= deppack;
+					// and it's transitive dependencies to current target
+					collectDependencies(deppack, ti, targets, level + 1);
+					continue;
+				}
+				auto depti = &targets[depname];
+				const depbs = &depti.buildSettings;
+				if (depbs.targetType == TargetType.executable)
+					continue;
+				// add to (link) dependencies
+				ti.dependencies ~= depname;
+				ti.linkDependencies ~= depname;
+
+				// recurse
+				collectDependencies(deppack, *depti, targets, level + 1);
+
+				// also recursively add all link dependencies of static libraries
+				// preserve topological sorting of dependencies for correct link order
+				if (depbs.targetType == TargetType.staticLibrary)
+					ti.linkDependencies = ti.linkDependencies.filter!(d => !depti.linkDependencies.canFind(d)).array ~ depti.linkDependencies;
+			}
 		}
 
-		// set pic for dynamic library builds.
-		if (buildsettings.targetType == TargetType.dynamicLibrary)
-			buildsettings.addOptions(BuildOption.pic);
+		collectDependencies(rootPackage, targets[rootPackage.name], targets);
+		static if (__VERSION__ > 2070)
+			visited.clear();
+		else
+			destroy(visited);
+
+		// 1. downwards inherits versions, debugVersions, and inheritable build settings
+		static void configureDependencies(in ref TargetInfo ti, TargetInfo[string] targets, size_t level = 0)
+		{
+			// do not use `visited` here as dependencies must inherit
+			// configurations from *all* of their parents
+			logDebug("%sConfigure dependencies of %s, deps:%(%s, %)", ' '.repeat(2 * level), ti.pack.name, ti.dependencies);
+			foreach (depname; ti.dependencies)
+			{
+				auto pti = &targets[depname];
+				mergeFromDependent(ti.buildSettings, pti.buildSettings);
+				configureDependencies(*pti, targets, level + 1);
+			}
+		}
 
-		logDiagnostic("Generate target %s (%s %s %s)", pack.name, buildsettings.targetType, buildsettings.targetPath, buildsettings.targetName);
-		if (is_target)
-			targets[pack.name] = TargetInfo(pack, [pack], configs[pack.name], buildsettings, null);
+		configureDependencies(targets[rootPackage.name], targets);
 
-		auto deps = pack.getDependencies(configs[pack.name]);
-		foreach (depname; deps.keys.sort()) {
-			auto depspec = deps[depname];
-			auto dep = m_project.getDependency(depname, depspec.optional);
-			if (!dep) continue;
+		// 2. add Have_dependency_xyz for all direct dependencies of a target
+		// (includes incorporated non-target dependencies and their dependencies)
```

Comment:
Since they are inherited upwards anyway, isn't it enough to set the `Have_xyz` constant for just `ti.packages`?